### PR TITLE
CLOUDP-114750: Add advanced cluster options

### PR DESCRIPTION
**Pull request by fabritsius**

### CODE REVIEW

File: test/int/dbuser_test.go [hunk 1]
```diff
@@ -148,27 +148,31 @@ var _ = Describe("AtlasDatabaseUser", Label("int", "AtlasDatabaseUser"), func()
...
+				func(g Gomega) {
+					success := testutil.WaitFor(k8sClient, createdClusterAzure, status.TrueCondition(status.ReadyType), validateClusterCreatingFuncGContext(g))()
+					g.Expect(success).To(BeTrue())
+				}).WithTimeout(ClusterUpdateTimeout).WithPolling(interval).Should(Succeed())
+		})
+	}
+
 	Describe("Create/Update two users, two clusters", func() {
 		It("They should be created successfully", func() {
-			By("Creating clusters", func() {
```
Comment (fabritsius) [inline test/int/dbuser_test.go:153]: The linter made me do this despite me not planning to change this file in any way.

File: pkg/controller/watch/event_handler.go [hunk 2]
```diff
@@ -2,10 +2,10 @@ package watch
 
 import (
 	"go.uber.org/zap"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
```
Comment (fabritsius) [inline pkg/controller/watch/event_handler.go:5]: Same thing as before, haven’t touched this file but linter demanded I sort these imports.

File: pkg/api/v1/atlascluster_types.go [hunk 3]
```diff
@@ -246,6 +252,43 @@ type ComputeSpec struct {
 	MaxInstanceSize string `json:"maxInstanceSize,omitempty"`
 }
 
+type ProcessArgs mongodbatlas.ProcessArgs
+
+func (specArgs ProcessArgs) IsEqual(newArgs interface{}) bool {
```
Comment (igor-karpukhin) [inline pkg/api/v1/atlascluster_types.go:257]: Good usage of reflection 👍

**Review by igor-karpukhin — APPROVED** [review]
Looks good

File: pkg/api/v1/atlascluster_types.go [hunk 4]
```diff
@@ -246,6 +252,43 @@ type ComputeSpec struct {
...
+	specV := reflect.ValueOf(specArgs)
+	newV := reflect.ValueOf(newArgs)
+	typeOfSpec := specV.Type()
+	for i := 0; i < specV.NumField(); i++ {
```
Comment (chatton) [inline pkg/api/v1/atlascluster_types.go:261]: do we have nested objects? Is it just top level keys we need to compare?

File: pkg/controller/atlascluster/atlascluster_controller.go [hunk 5]
```diff
@@ -114,28 +112,41 @@ func (r *AtlasClusterReconciler) Reconcile(context context.Context, req ctrl.Req
...
+	handleCluster := r.selectClusterHandler(cluster)
+	if result, _ := handleCluster(ctx, project, cluster); !result.IsOk() {
+		ctx.SetConditionFromResult(status.ClusterReadyType, result)
+		return result.ReconcileResult(), nil
+	}
+
+	if result := r.handleAdvancedOptions(ctx, project, cluster); !result.IsOk() {
+		ctx.SetConditionFromResult(status.ClusterReadyType, result)
+		return result.ReconcileResult(), nil
+	}
```
Comment (chatton) [inline pkg/controller/atlascluster/atlascluster_controller.go:124]: do we always want to handle advanced options? How does this behave when we are dealing with non-advanced clusters?

**Review by chatton — COMMENTED** [review]
Looking good! Just had a few questions

File: pkg/api/v1/atlascluster_types.go (hunk 4 above)
Comment (fabritsius) [reply pkg/api/v1/atlascluster_types.go:261]: Yeah, thankfully no nested objects in that case, so works as is 😌

File: pkg/controller/atlascluster/atlascluster_controller.go (hunk 5 above)
Comment (fabritsius) [reply pkg/controller/atlascluster/atlascluster_controller.go:124]: Advanced options are not related to the Advanced Clusters and can be set/changed for both cluster types.

**Review by chatton — APPROVED** [review]
LGTM! The changes all look good 🥇

**Pull request merged**